_posts/2023-09-19-virtual-thread-1.adoc

```diff
@@ -0,0 +1,335 @@
+
+    // `service` is a rest client, it executes an I/O operation
+    // but this time it returns a Uni<String>, so it's not blocking
+    return service.asyncGreetings() // Non-blocking
+        .map(resp -> {
+          // This is the continuation, the code executed once the
+          // response is received
+          return resp.toUpperCase();
+        });
+    }
+}
+----
+
+An application with this code handles more concurrent requests and uses less memory than the imperative one, but, the development model is different.
+
+Most of the time, the reactive and imperative models are opposed.
+This does not need to be the case.
+Quarkus uses a reactive core and lets you decide if you want to use the reactive or imperative model. 
+Check the https://quarkus.io/blog/resteasy-reactive-smart-dispatch/['to block or not to block' article] for more details about this ability.
+
+## What do virtual threads change?
+
+Java 19 introduced a new type of thread: virtual threads.
+In Java 21, this API became generally available.
+
+But what are these virtual threads?
+Virtual threads reuse the idea of the reactive paradigm but allow an imperative development model.
+You get the benefits from the reactive and imperative models without the drawbacks! 
+
+Like with green threads, virtual threads are managed by the JVM. 
+Virtual threads occupy less space than platform threads in memory. 
+Hence, using more virtual threads than platform threads simultaneously becomes possible without blowing up the memory.
+Virtual threads are supposed to be disposable entities that we create when we need them; pooling or reusing them for different tasks is discouraged.
+
+But what does it change?
+Blocking a virtual thread is, in general, very cheap! 
+There is a pinch of magic that makes virtual thread very appealing.
+When your code running on a virtual thread needs to execute an I/O operation, it uses a blocking API.
+So, the code waits for the result, as with the imperative model.
+However, since the JVM manages virtual threads, no underlying OS thread is blocked when they perform this blocking operation. 
+The state of the virtual thread is stored in the heap, and another virtual thread can be executed on the same Java platform (carrier) thread, exactly as in the reactive model.
+When the I/O operation completes, the virtual thread becomes executable again, and when a carrier thread is available, the state of the virtual thread is restored, and the execution continues.
+For the developer, all this mechanic is invisible! 
+You just write synchronous code, and it's executed like proper reactive code without blocking the OS thread.
+
+Your code runs on top of virtual threads, but under the hood, only a few carrier threads execute them. 
+
+To summarize, virtual threads are:
+
+    1. Lightweight - you can have a LOT of them
+    2. Cheap to create - no need to pool them anymore
+    3. Cheap to block - blocking a virtual thread does not block the underlying OS thread
```

Comment:
should this be "cheap to block using i/o" as I assume blocking using raw CPU is still expensive.